pkg/cache/nodes.go

```diff
@@ -63,15 +109,21 @@ func (nc *NodeController) addNode(obj interface{}) {
 	log.Logger.Info("adding node to cache", zap.String("NodeName", node.Name))
 	nc.cache.AddNode(node)
 
-	n := common.CreateFrom(node)
-	request := common.CreateUpdateRequestForNewNode(n)
-	log.Logger.Info("report new nodes to scheduler", zap.Any("request", request.String()))
-	if err := nc.proxy.Update(&request); err != nil {
-		log.Logger.Error("hitting error while handling AddNode", zap.Error(err))
-	}
+	// add node to nodes map
+	newNode := newSchedulerNode(node.Name, string(node.UID), common.GetNodeResource(&node.Status), nc.proxy)
+	nc.nodesMap[node.Name] = newNode
+
+	// once node is added to scheduler, first thing is to recover its state
+	dispatcher.Dispatch(CachedSchedulerNodeEvent{
+		NodeId:    newNode.name,
+		Event:     events.RecoverNode,
+	})
 }
 
-func (nc *NodeController) updateNode(oldObj, newObj interface{}) {
+func (nc *schedulerNodes) updateNode(oldObj, newObj interface{}) {
+	nc.lock.Lock()
```

Comment:
Can we move the locking down to the point were we need it to make the lock cycle as short as possible?
Just before we call UpdateNode()

Reply:
Yes, we can, but I don't think it's necessary. These few lines are not time-consuming operations, it can hardly impact the performance. But having them at the beginning, IMO it is more readable. When we do performance evaluations, I expect there will be a ton of places that can be optimized and such places can be ignored. Would that be OK?